### PR TITLE
Add gripper calibration module

### PR DESCRIPTION
**Pull request by pinakjani**
Added Gripper Calibration script to perform a calibration of gripper zero value. The script subscribes to the joint states positions and calculates the error. Upon reaching ~zero error it notes the offset value and passes it to the SDK via gripper_calib_srv (ROS Service). This service updates the gripper offset in the joint states using the gripper_map. 

To make this process modular the Gripper Calibration routine takes in parameters for the robot name and the motor config path. This node is added to the xsarm_control.launch in the interbotix_ros_manipulators package. Thus upon startup of the controller launch file, the gripper_calibration node is executed.

I have made a pull request for changes in the launch and yaml files in the interbotix_ros_manipulators package to make it compatible with all the robots.

### CODE REVIEW

File: interbotix_ros_xseries/interbotix_xs_sdk/include/interbotix_xs_sdk/xs_sdk_obj.h [hunk 1]
```diff
@@ -55,7 +57,8 @@ struct Gripper                                                          // Struc
   float horn_radius;                                                    // Distance [m] from the motor horn's center to its edge
   float arm_length;                                                     // Distance [m] from the edge of the motor horn to a finger
   std::string left_finger;                                              // Name of the 'left_finger' joint as defined in the URDF (if present)
-  std::string right_finger;                                             // Name of the 'right_finger' joint as defined in the URDF (if present)
+  std::string right_finger;
```
Comment (LSinterbotix) [inline interbotix_ros_xseries/interbotix_xs_sdk/include/interbotix_xs_sdk/xs_sdk_obj.h:60]: Undo removing the right_finger description comment.

File: interbotix_ros_xseries/interbotix_xs_sdk/include/interbotix_xs_sdk/xs_sdk_obj.h [hunk 2]
```diff
@@ -55,7 +57,8 @@ struct Gripper                                                          // Struc
   float horn_radius;                                                    // Distance [m] from the motor horn's center to its edge
   float arm_length;                                                     // Distance [m] from the edge of the motor horn to a finger
   std::string left_finger;                                              // Name of the 'left_finger' joint as defined in the URDF (if present)
-  std::string right_finger;                                             // Name of the 'right_finger' joint as defined in the URDF (if present)
+  std::string right_finger; 
+  float calibration_offset;                                             // Calibration offset fro gripper in angles
```
Comment (LSinterbotix) [inline interbotix_ros_xseries/interbotix_xs_sdk/include/interbotix_xs_sdk/xs_sdk_obj.h:61]:
```suggestion
  float calibration_offset;                                             // Calibration offset [rad] for gripper joint
```

File: interbotix_ros_xseries/interbotix_xs_sdk/src/xs_sdk_obj.cpp [hunk 3]
```diff
@@ -1259,6 +1277,9 @@ void InterbotixRobotXS::robot_update_joint_states(const ros::TimerEvent &e)
   }
   // Publish the message to the joint_states topic
   joint_state_msg.header.stamp = ros::Time::now();
+  for(const auto gripper:gripper_map){
+      joint_state_msg.position[gripper.second.js_index]-= gripper.second.calibration_offset;
+  }
```
Comment (LSinterbotix) [inline interbotix_ros_xseries/interbotix_xs_sdk/src/xs_sdk_obj.cpp:1282]:
```suggestion
  for (const auto gripper : gripper_map){
      joint_state_msg.position[gripper.second.js_index] -= gripper.second.calibration_offset;
  }
```

File: interbotix_ros_xseries/interbotix_xs_sdk/src/xs_sdk_obj.cpp [hunk 4]
```diff
@@ -1086,6 +1088,22 @@ bool InterbotixRobotXS::robot_srv_get_motor_registers(interbotix_xs_msgs::Regist
...
 
+/**
+ * @brief ROS Service to set the gripper calibration value in calibration map
+ *
+ * @param req
+ * @param res
+ * @return true
+ * @return false
+ * @details - refer to the service definition for details
+ */
```
Comment (LSinterbotix) [inline interbotix_ros_xseries/interbotix_xs_sdk/src/xs_sdk_obj.cpp:1099]: Add function docs, match style of other function docs

File: interbotix_ros_xseries/interbotix_xs_sdk/include/interbotix_xs_sdk/xs_sdk_obj.h [hunk 5]
```diff
@@ -187,10 +190,12 @@ class InterbotixRobotXS
   ros::ServiceServer srv_get_robot_info;                                        // ROS Service Server used to get general information about the robot (like limits)
   ros::ServiceServer srv_torque_enable;                                         // ROS Service Server used to torque on/off any motor
   ros::ServiceServer srv_reboot_motors;                                         // ROS Service Server used to reboot any motor
+  ros::ServiceServer srv_gripper_calib;                                         // ROS Service Server used to get the gripper calibration value
   ros::Timer tmr_joint_states;                                                  // ROS Timer used to continuously publish joint states
   ros::Timer tmr_joint_traj;                                                    // ROS One-Shot Timer used when commanding motor trajectories
   sensor_msgs::JointState joint_states;                                         // Holds the most recent JointState message
   interbotix_xs_msgs::JointTrajectoryCommand joint_traj_cmd;                     // Holds the most recent JointTrajectoryCommand message
+  float prev_gripper_pos = 0;
```
Comment (LSinterbotix) [inline interbotix_ros_xseries/interbotix_xs_sdk/include/interbotix_xs_sdk/xs_sdk_obj.h:198]: Is this used?

File: interbotix_ros_xseries/interbotix_xs_sdk/include/interbotix_xs_sdk/xs_sdk_obj.h [hunk 6]
```diff
@@ -56,6 +58,7 @@ struct Gripper                                                          // Struc
   float arm_length;                                                     // Distance [m] from the edge of the motor horn to a finger
   std::string left_finger;                                              // Name of the 'left_finger' joint as defined in the URDF (if present)
   std::string right_finger;                                             // Name of the 'right_finger' joint as defined in the URDF (if present)
+  float calibration_offset;                                             // Calibration offset in [rad] for gripper
```
Comment (LSinterbotix) [inline interbotix_ros_xseries/interbotix_xs_sdk/include/interbotix_xs_sdk/xs_sdk_obj.h:61]:
```suggestion
  float calibration_offset;                                             // Calibration offset [rad] for gripper
```

File: interbotix_ros_xseries/interbotix_xs_sdk/include/interbotix_xs_sdk/xs_sdk_obj.h [hunk 7]
```diff
@@ -309,6 +314,17 @@ class InterbotixRobotXS
...
 
+
+  /**
+   * @brief Sets the gripper calibration offset to gripper map.
+   *
+   * @param req  - Consists the gripper name and its position offset value
+   * @param res  [out] - Calibration offset service message response [unused]
+   * @return true
+   * @return false
+   */
```
Comment (LSinterbotix) [inline interbotix_ros_xseries/interbotix_xs_sdk/include/interbotix_xs_sdk/xs_sdk_obj.h:325]: Add return value docs, match style of other function docs

**Review by LSinterbotix — CHANGES_REQUESTED** [review]

File: interbotix_ros_xseries/interbotix_xs_sdk/src/xs_sdk_obj.cpp (hunk 4 above)
Comment (pinakjani) [reply interbotix_ros_xseries/interbotix_xs_sdk/src/xs_sdk_obj.cpp:1099]: Added

File: interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+int main( int argc, char** argv )
+{
+    ros::init(argc, argv, "gripper_calib");
+    std::string param1,param2,motor_configs_file;
+    ros::NodeHandle n1;
+    ros::param::get("~motor_configs", motor_configs_file);
+    ros::param::get("~robot", param1);
+    ROS_INFO("Got parameter 1 : %s", param1.c_str());
+    ROS_INFO("Got parameter 2 : %s",motor_configs_file.c_str());
```
Comment (LSinterbotix) [inline interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp:128]: Name variables better and log them more descriptively.

File: interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+        ROS_FATAL("[xs_sdk] YAML Error: '%s'", error.what());
+        success = false;
+    }
+    if (yaml_node.IsNull())
+    {
+        ROS_FATAL("[xs_sdk] Calibration Config file was not found. Shutting down...");
+        success = false;
+    }
+    YAML::Node all_grippers = yaml_node["grippers"];
+    for(const auto gripper:all_grippers){
```
Comment (LSinterbotix) [inline interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp:112]: Spaces between operators. Loop braces on next line
```suggestion
    for(const auto gripper : all_grippers)
    {
```

File: interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+    }
+    return calibration_joints;
+
+}
+
+
+int main( int argc, char** argv )
+{
+    ros::init(argc, argv, "gripper_calib");
+    std::string param1,param2,motor_configs_file;
```
Comment (LSinterbotix) [inline interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp:123]: ```param2``` is unused

File: interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+        Gripper_calibration obj(&n1,success,param1,name);
+        ros::Subscriber sub = n1.subscribe("/"+param1+"/joint_states",1000,&Gripper_calibration::calib_callback,&obj);
+        if (success)
+            ros::spin();
+        else
+        {
+            ROS_FATAL(
+                "[xs_sdk] For troubleshooting, please see "
+                "'https://docs.trossenrobotics.com/interbotix_xsarms_docs/troubleshooting.html'");
+            // ros::shutdown();
```
Comment (LSinterbotix) [inline interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp:142]: Remove or implement commented line

File: interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+
+    /**
+     * @brief Subscriber callback for performing gripper calibration
+     */
+    void calib_callback(const sensor_msgs::JointState &msg){
+        static int count = 0;
+        if(count%40==0){
+            int index = find_gripper_index(msg.name);
+            float curr_gripper_pos = msg.position[index];
+            ROS_INFO("I heard this position [%f]",msg.position[index]);
```
Comment (LSinterbotix) [inline interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp:50]: Some of these ROS_INFO logging lines read like leftover debug messages for development. This one can probably be removed.

File: interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+     */
+    void calib_complete_srv(float min_position_offset, std::string _gripper_name){
+        gripper_calib_srv.request.gripper_name = _gripper_name;
+        gripper_calib_srv.request.offset = min_position_offset;
+        bool complete = client.call(gripper_calib_srv);
+        if(complete){
+            ROS_INFO("Gripper calibration finished successfully");
+        }
+        else{
+            ROS_ERROR("Failed to call service torque_enable");
```
Comment (LSinterbotix) [inline interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp:79]: We're calling the calibration service here, not torque_enable. Should we do any error handling here?

**Review by LSinterbotix — CHANGES_REQUESTED** [review]

File: interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp (hunk 10 above)
Comment (pinakjani) [reply interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp:125]: Removing that

File: interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+class Gripper_calibration{
+private:
+    ros::NodeHandle node;
+    ros::Publisher pub;
+    ros::ServiceClient client;
+    ros::Subscriber sub;
+    float prev_val = 0.0;
+    interbotix_xs_msgs::JointSingleCommand gripper_command_msg;
+    interbotix_xs_msgs::GripperCalib gripper_calib_srv;
+    std::string _gripper_name;
```
[the remark this answers is not in the document]
Comment (pinakjani) [reply interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp:17]: Added

File: interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+public:
+    /**
+     * @brief Construct a new Gripper_calibration object.
+     * @param node_handle - Node object
+     * @param success  - Checker flag if node processes are succesful
+     * @param robot_name  - Name of the robot
+     * @param gripper_name  - Name of the gripper
+     */
```
[the remark this answers is not in the document]
Comment (pinakjani) [reply interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp:25]: Fixing this, However the use /***/ seems the standard way with DOXYGEN, I use the doxygen documentation extension that makes this commenting easier.

File: interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+    Gripper_calibration(ros::NodeHandle* node_handle, bool &success, std::string robot_name, std::string gripper_name):node(*node_handle),_gripper_name(gripper_name)
```
[the remark this answers is not in the document]
Comment (pinakjani) [reply interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp:26]: Done

File: interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+    {
+        pub = node.advertise<interbotix_xs_msgs::JointSingleCommand>("/"+robot_name+"/commands/joint_single",1000);
```
[the remark this answers is not in the document]
Comment (pinakjani) [reply interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp:28]: Fixed

File: interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+        client = node.serviceClient<interbotix_xs_msgs::GripperCalib>("/"+robot_name+"/gripper_calibration");
+        calib_complete_srv(0.0,_gripper_name);
+    }
+
+    /**
+     * @brief Returns the gripper index in joint states of the parsed gripper name
+     * @param names - list of joint names
+     * @return int  - index of the gripper name
+     */
+    int find_gripper_index(const std::vector<std::string>& names){
```
[the remark this answers is not in the document]
Comment (pinakjani) [reply interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp:39]: Done

File: interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+        return std::find(names.begin(),names.end(),_gripper_name)-names.begin();
```
[the remark this answers is not in the document]
Comment (pinakjani) [reply interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp:40]: Fixed

File: interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp (hunk 12 above)
Comment (pinakjani) [reply interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp:51]: Removed

File: interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+        }
+        count++;
+    }
+
+    /**
+     * @brief ROS Service Client that sends the calibration offset values to SDK
+     * @param min_position_offset - gripper offset value
+     * @param _gripper_name - gripper name
+     */
+    void calib_complete_srv(float min_position_offset, std::string _gripper_name){
```
[the remark this answers is not in the document]
Comment (pinakjani) [reply interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp:72]: Fixed it

File: interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+        gripper_calib_srv.request.gripper_name = _gripper_name;
+        gripper_calib_srv.request.offset = min_position_offset;
+        bool complete = client.call(gripper_calib_srv);
+        if(complete){
```
[the remark this answers is not in the document]
Comment (pinakjani) [reply interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp:76]: Not sure of when to use cuddle braces and when not. Is this consistent with the entire code structure? It seems different with what ROS2 linting follows.

File: interbotix_ros_xseries/interbotix_xs_sdk/include/interbotix_xs_sdk/xs_sdk_obj.h [hunk 22]
```diff
@@ -309,6 +312,16 @@ class InterbotixRobotXS
...
   bool robot_srv_validate(const std::string &cmd_type, std::string &name);
 
+
+  /**
+   * @brief Sets the gripper calibration offset to gripper map.
+   * @param req  - Consists the gripper name and its position offset value
+   * @param res  [out] - Calibration offset service message response [unused]
+   * @return true if the service call request is valid, false otherwise
+   * @details - refer to the service definition for details
+   */
```
[the remark this answers is not in the document]
Comment (pinakjani) [reply interbotix_ros_xseries/interbotix_xs_sdk/include/interbotix_xs_sdk/xs_sdk_obj.h]: Resolved

File: interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp [hunk 23]
```diff
@@ -0,0 +1,141 @@
...
+class GripperCalibration
+{
+public:
+
+    /// @brief Construct a new Gripper_calibration object.
+    /// @param node_handle - Node object
+    /// @param success  - Checker flag if node processes are succesful
+    /// @param robot_name  - Name of the robot
+    /// @param gripper_name  - Name of the gripper
+    GripperCalibration(ros::NodeHandle* node_handle, bool &success, std::string robot_name, std::string gripper_name)
```
Comment (LSinterbotix) [inline interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp:15]: Remove success

File: interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp [hunk 24]
```diff
@@ -0,0 +1,141 @@
...
+    ros::param::get("~motor_configs", motor_configs_file);
+    ros::param::get("~robot", robot_name);
+    ROS_INFO("Got parameter 1 : %s", robot_name.c_str());
+    ROS_INFO("Got parameter 2 : %s",motor_configs_file.c_str());
+    bool success = true;
+    std::vector<std::string> gripper_order = load_calibration_config(success, robot_name, motor_configs_file);
+    for(const auto gripper_name : gripper_order) {
+        ROS_INFO("%s",gripper_name.c_str());
+        GripperCalibration obj(&n1, success, robot_name, gripper_name);
+        ros::Subscriber sub = n1.subscribe("/"+robot_name+"/joint_states",1000,&GripperCalibration::calib_callback,&obj);
```
Comment (LSinterbotix) [inline interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp:129]:
```suggestion
        ros::Subscriber sub = n1.subscribe("joint_states", 1000, &GripperCalibration::calib_callback, &obj);
```

File: interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp [hunk 25]
```diff
@@ -0,0 +1,141 @@
...
+        if (success)
+            ros::spin();
+        else
+        {
+            ROS_FATAL(
+                "[xs_sdk] For troubleshooting, please see "
```
Comment (LSinterbotix) [inline interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp:135]:
```suggestion
                "[gripper_calib] For troubleshooting, please see "
```

File: interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp [hunk 26]
```diff
@@ -0,0 +1,141 @@
...
+    std::string robot_name, motor_configs_file;
+    ros::NodeHandle n1;
+    ros::param::get("~motor_configs", motor_configs_file);
+    ros::param::get("~robot", robot_name);
+    ROS_INFO("Got parameter 1 : %s", robot_name.c_str());
+    ROS_INFO("Got parameter 2 : %s",motor_configs_file.c_str());
+    bool success = true;
+    std::vector<std::string> gripper_order = load_calibration_config(success, robot_name, motor_configs_file);
+    for(const auto gripper_name : gripper_order) {
+        ROS_INFO("%s",gripper_name.c_str());
```
Comment (LSinterbotix) [inline interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp:127]: Check if any and all logging lines actually need to be there

File: interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp [hunk 27]
```diff
@@ -0,0 +1,141 @@
...
+        ROS_FATAL("[gripper_calib] Calibration Config file was not found or has a bad format. Shutting down...");
+        success = false;
+    }
+    YAML::Node all_grippers = yaml_node["grippers"];
+    for(const auto gripper : all_grippers)
+    {
+        calibration_joints.push_back(gripper.first.as<std::string>());
+    }
+    return calibration_joints;
+
```
Comment (LSinterbotix) [inline interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp:111]:
```suggestion
```

File: interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp [hunk 28]
```diff
@@ -0,0 +1,141 @@
...
+    /// @brief Returns the gripper index in joint states of the parsed gripper name
+    /// @param names - list of joint names
+    /// @return int  - index of the gripper name
+    int find_gripper_index(const std::vector<std::string>& names)
+    {
+        return std::find(names.begin(), names.end(), _gripper_name) - names.begin();
+    }
+
+
```
Comment (LSinterbotix) [inline interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp:63]:
```suggestion
```

File: interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp [hunk 29]
```diff
@@ -0,0 +1,141 @@
...
+        if(count%40==0){
+            int index = find_gripper_index(msg.name);
+            float curr_gripper_pos = msg.position[index];
+            if(std::abs(curr_gripper_pos-prev_val)>0.0001){
+                gripper_command_msg.name = _gripper_name;
+                gripper_command_msg.cmd = -200;
+                pub.publish(gripper_command_msg);
+            }
+            else{
+                ROS_INFO("The error is 0");
```
Comment (LSinterbotix) [inline interbotix_ros_xseries/interbotix_xs_sdk/src/gripper_calib.cpp:36]: Remove?

**Review by LSinterbotix — CHANGES_REQUESTED** [review]

**Review by LSinterbotix — APPROVED** [review]

**Pull request merged**